matrix/matrix.py

```diff
@@ -0,0 +1,146 @@
+        return self._matrix
+
+    def _add_matrix(self, other):
+        """Add two matrices.
+
+        Args:
+            other (Matrix): the other matrix to operate on.
+
+        Returns:
+            Matrix: the result of the addition.
+        """
+        dimension = len(self.tuples)
+
+        if len(other.tuples) != dimension:
+            raise ValueError("Matrices length don't match")
+
+        calculated_data = [[0 for i in range(dimension)]
+                           for j in range(dimension)]
+
+        for i in range(len(self.tuples)):
+            for j in range(len(self.tuples)):
+                calculated_data[i][j] = self.tuples[i][j] + other.tuples[i][j]
+
+        calculated_data = [tuple(line) for line in calculated_data]
+        return Matrix(tuple(calculated_data))
+
+    def _matrix_mul_matrix(self, other):
+        """Multiply two matrices.
+
+        Args:
+            other (Matrix): the other matrix to multiply.
+
+        Returns:
+            Matrix: the product.
+        """
+        dimension = len(self.tuples)
+
+        if len(other.tuples) != dimension:
+            raise ValueError(
+                f"Dimension {len(other.tuples)} don't match {dimension}")
+
+        calculated_data = [[0 for i in range(dimension)]
+                           for j in range(dimension)]
+
+        for i in range(dimension):
+            for k in range(dimension):
+                for j in range(dimension):
+                    calculated_data[i][j] += self.tuples[i][k] * \
+                                             other.tuples[k][j]
+
+        calculated_data = [tuple(line) for line in calculated_data]
+        return Matrix(tuple(calculated_data))
+
+    def __add__(self, other):
+        """Add two matrices."""
+        return self._add_matrix(other)
+
+    def __sub__(self, other):
+        """Subtract two matrices."""
+        return self + (-1 * other)
+
+    def __mul__(self, other):
+        """Multiply two matrices or scalar by matrix."""
+        if isinstance(other, Matrix):
+            return self._matrix_mul_matrix(other)
+
+        # if scalar multiplication
+        return Matrix(tuple(
+            [tuple(map(lambda x: x * other, line)) for line in self.tuples]))
+
+    def __rmul__(self, other):
+        """Define reverse multiplication as regular multiplication."""
+        return self * other
+
+    def __truediv__(self, other):
+        """Divide matrix by scalar. Return floats if needed."""
+        return self * (1.0 / other)
+
+    def __eq__(self, other):
+        """Compare two matrices. Return true if they are equal else false."""
+        return self.tuples == other.tuples
+
+    def __ne__(self, other):
+        """Compare two matrices. Return true if they are not equal else
+        false."""
+        return not self == other
+
+    def __iter__(self):
+        """Enable to iterate on a matrix lines'."""
+        return iter(self.tuples)
```

Comment:
tuples are already iterable

Reply:
If I won't put in `iter()` it will raise: `TypeError: iter() returned non-iterator of type 'tuple'`

Reply:
you are right, my mistake.
__iter__ expects iterator, not iterable